File: src/kyber_pqc/pem.py
"""Hex-encoded PEM serialization for Kyber-512 material."""

from __future__ import annotations

import re
from enum import Enum
from pathlib import Path
from typing import Union

from .core import (
    CIPHERTEXT_BYTES,
    PRIVATE_KEY_BYTES,
    PUBLIC_KEY_BYTES,
    SHARED_SECRET_BYTES,
)

PEM_LINE_WIDTH = 64


class PEMKind(str, Enum):
    PUBLIC_KEY = "PUBLIC KEY"
    PRIVATE_KEY = "PRIVATE KEY"
    CIPHERTEXT = "CIPHERTEXT"
    SHARED_SECRET = "SHARED SECRET"


_EXPECTED_BYTES = {
    PEMKind.PUBLIC_KEY: PUBLIC_KEY_BYTES,
    PEMKind.PRIVATE_KEY: PRIVATE_KEY_BYTES,
    PEMKind.CIPHERTEXT: CIPHERTEXT_BYTES,
    PEMKind.SHARED_SECRET: SHARED_SECRET_BYTES,
}


def _header(kind: PEMKind) -> str:
    return f"-----BEGIN KYBER512 {kind.value}-----"


def _footer(kind: PEMKind) -> str:
    return f"-----END KYBER512 {kind.value}-----"


def encode_pem(kind: PEMKind, data: bytes) -> str:
    """Encode binary Kyber material as a hex PEM block."""
    expected = _EXPECTED_BYTES[kind]
    if len(data) != expected:
        raise ValueError(
            f"invalid data length for {kind.value}: "
            f"expected {expected}, got {len(data)}"
        )

    hex_body = data.hex()
    lines = [
        hex_body[i: i + PEM_LINE_WIDTH]
        for i in range(0, len(hex_body), PEM_LINE_WIDTH)
    ]
    return "\n".join([_header(kind), *lines, _footer(kind), ""])


def decode_pem(kind: PEMKind, pem: str) -> bytes:
    """Decode a hex PEM block back to binary Kyber material."""
    expected = _EXPECTED_BYTES[kind]
    begin = re.escape(_header(kind))
    end = re.escape(_footer(kind))
    match = re.search(
        rf"{begin}\s*([0-9a-fA-F\s]+?)\s*{end}",
        pem,
        flags=re.DOTALL,
    )
    if match is None:
        raise ValueError(f"missing or invalid PEM block for {kind.value}")

    body = re.sub(r"\s+", "", match.group(1))
    try:
        decoded = bytes.fromhex(body)
    except ValueError as exc:
        raise ValueError(f"invalid hex payload for {kind.value}") from exc

    if len(decoded) != expected:
        raise ValueError(
            f"invalid decoded length for {kind.value}: "
            f"expected {expected}, got {len(decoded)}"
        )
    return decoded


def write_pem_file(
    path: Union[str, Path],
    kind: PEMKind,
    data: bytes,
) -> None:
    """Write a hex PEM file with restrictive permissions."""
    pem_path = Path(path)
    pem_path.write_text(encode_pem(kind, data), encoding="utf-8")
    pem_path.chmod(0o600)


def read_pem_file(path: Union[str, Path], kind: PEMKind) -> bytes:
    """Read binary Kyber material from a hex PEM file."""
    return decode_pem(kind, Path(path).read_text(encoding="utf-8"))

File: go/kyberpqc/pem.go
package kyberpqc

import (
	"encoding/hex"
	"fmt"
	"os"
	"strings"
)

const pemLineWidth = 64

type PEMKind int

const (
	PEMPublicKey PEMKind = iota
	PEMPrivateKey
	PEMCiphertext
	PEMSharedSecret
)

var pemLabels = map[PEMKind]string{
	PEMPublicKey:   "PUBLIC KEY",
	PEMPrivateKey:  "PRIVATE KEY",
	PEMCiphertext:  "CIPHERTEXT",
	PEMSharedSecret: "SHARED SECRET",
}

var pemExpectedBytes = map[PEMKind]int{
	PEMPublicKey:   PublicKeyBytes,
	PEMPrivateKey:  PrivateKeyBytes,
	PEMCiphertext:  CiphertextBytes,
	PEMSharedSecret: SharedSecretBytes,
}

func PEMLabel(kind PEMKind) (string, error) {
	label, ok := pemLabels[kind]
	if !ok {
		return "", fmt.Errorf("unknown pem kind: %d", kind)
	}
	return label, nil
}

func EncodePEM(kind PEMKind, data []byte) (string, error) {
	expected, ok := pemExpectedBytes[kind]
	if !ok {
		return "", fmt.Errorf("unknown pem kind: %d", kind)
	}
	if len(data) != expected {
		return "", fmt.Errorf("invalid data length: expected %d, got %d", expected, len(data))
	}

	label, err := PEMLabel(kind)
	if err != nil {
		return "", err
	}

	encoded := hex.EncodeToString(data)
	var builder strings.Builder
	builder.WriteString(fmt.Sprintf("-----BEGIN KYBER512 %s-----\n", label))
	for i := 0; i < len(encoded); i += pemLineWidth {
		end := i + pemLineWidth
		if end > len(encoded) {
			end = len(encoded)
		}
		builder.WriteString(encoded[i:end])
		builder.WriteByte('\n')
	}
	builder.WriteString(fmt.Sprintf("-----END KYBER512 %s-----\n", label))
	return builder.String(), nil
}

func DecodePEM(kind PEMKind, pem string) ([]byte, error) {
	label, err := PEMLabel(kind)
	if err != nil {
		return nil, err
	}
	expected, ok := pemExpectedBytes[kind]
	if !ok {
		return nil, fmt.Errorf("unknown pem kind: %d", kind)
	}

	begin := fmt.Sprintf("-----BEGIN KYBER512 %s-----", label)
	end := fmt.Sprintf("-----END KYBER512 %s-----", label)

	start := strings.Index(pem, begin)
	if start < 0 {
		return nil, fmt.Errorf("missing pem header for %s", label)
	}
	start += len(begin)
	stop := strings.Index(pem[start:], end)
	if stop < 0 {
		return nil, fmt.Errorf("missing pem footer for %s", label)
	}

	body := strings.Map(func(r rune) rune {
		switch r {
		case '\n', '\r', ' ':
			return -1
		default:
			return r
		}
	}, pem[start:start+stop])

	decoded, err := hex.DecodeString(body)
	if err != nil {
		return nil, err
	}
	if len(decoded) != expected {
		return nil, fmt.Errorf("invalid decoded length: expected %d, got %d", expected, len(decoded))
	}
	return decoded, nil
}

func WritePEMFile(path string, kind PEMKind, data []byte) error {
	pem, err := EncodePEM(kind, data)
	if err != nil {
		return err
	}
	return os.WriteFile(path, []byte(pem), 0o600)
}

func ReadPEMFile(path string, kind PEMKind) ([]byte, error) {
	content, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	return DecodePEM(kind, string(content))
}
